### PR TITLE
Evaluate check constraints during merge

### PR DESCRIPTION
**Pull request by fulghum**
Adds support for evaluating check constraint expressions on a table during a merge. 

Fixes https://github.com/dolthub/dolt/issues/3462

### CODE REVIEW

File: go/libraries/doltcore/merge/merge_prolly_rows.go [hunk 1]
```diff
@@ -856,6 +1040,41 @@ func (m *secondaryMerger) finalize(ctx context.Context) (durable.IndexSet, durab
...
+	mockProvider := memory.NewDBProvider(mockDatabase)
+	catalog := analyzer.NewCatalog(mockProvider)
+
+	pseudoAnalyzedQuery, err := planbuilder.Parse(ctx, catalog, query)
+	if err != nil {
+		return nil, err
+	}
+
+	var expr sql.Expression
+	transform.Inspect(pseudoAnalyzedQuery, func(n sql.Node) bool {
```
Comment (max-hoffman) [inline go/libraries/doltcore/merge/merge_prolly_rows.go:1064]: is it possible for the expression to have a nested projection?

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_merge.go [hunk 2]
```diff
@@ -4318,6 +4321,210 @@ var ThreeWayMergeWithSchemaChangeTestScripts = []MergeScriptTest{
...
+		RightSetUpScript: []string{
+			"update t set col2=4;",
+		},
+		LeftSetUpScript: []string{
+			"update t set col1=4;",
+		},
+		Assertions: []queries.ScriptTestAssertion{
+			{
+				Query:    "call dolt_merge('right');",
+				Expected: []sql.Row{{0, 0x1}},
```
Comment (max-hoffman) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries_merge.go:4342]: any reason you use hex for these?

**Review by max-hoffman — APPROVED** [review]
LGTM! excited to see the name resolving in action

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_merge.go (hunk 2 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries_merge.go]: good question... looks like that was just getting copied around. I just pushed 33755ab12bcdbd6a1b4575ad2884c3d2558990a9 that cleans that up.

File: go/libraries/doltcore/merge/merge_prolly_rows.go [hunk 3]
```diff
@@ -335,6 +351,175 @@ func threeWayDiffer(ctx context.Context, tm *TableMerger, valueMerger *valueMerg
...
+	if err != nil {
+		return err
+	}
+	cvm := prolly.ConstraintViolationMeta{VInfo: vinfo, Value: value}
+	return cv.edits.ReplaceConstraintViolation(ctx, key, cv.srcHash, prolly.ArtifactTypeChkConsViol, cvm)
+}
+
+// buildRow takes the |key| and |value| tuple and returns a new sql.Row, along with any errors encountered.
+func (cv checkValidator) buildRow(ctx *sql.Context, key, value val.Tuple) (sql.Row, error) {
+	// When we parse and resolve the check constraint expression with planbuilder, it leaves row position 0
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/merge/merge_prolly_rows.go:486]: Ahh, thanks for the extra details here and for pointing out `FixFieldIndexesForExpressions`. Check constraint expressions are pretty limited and can't reference other tables, so sounds like this should be sufficient for this code.

**Pull request merged**